### PR TITLE
Added tesnorizeation for avx2 based gemm.

### PR DESCRIPTION
**Pull request by kimishpatel**
Summary:
Similar to other avx512 tensorization that reduces data:1x4 and kernel:16x4 to output:1x16, this PR introduces similar reduction using avx2 tensorization. It keeps the same API as avx512 so as to not have to introduce a new memory layout for weights.

Test Plan:
on avx2 machine:
python tests/python/contrib/test_gemm_avx2_acc32.py

Reviewers:

Subscribers:

Tasks:

Tags:

Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

### CODE REVIEW

**Comment by kimishpatel** [comment]
Depends on this PR: https://github.com/dmlc/tvm/pull/3981

File: topi/python/topi/x86/tensor_intrin.py [hunk 1]
```diff
@@ -275,3 +275,97 @@ def _instr(index):
...
+
+            a_int8 = ins[0].vload([0], "uint8x4")
+            re_int32 = tvm.call_pure_intrin('int32', 'reinterpret', a_int8)
+            vec_ai32 = re_int32.astype('int32x8')
+            vec_a = tvm.call_pure_intrin('int8x32', 'reinterpret', vec_ai32)
+            vec_b_0 = ins[1].vload([0, 0], "int8x32")
+            vec_b_1 = ins[1].vload([8, 0], "int8x32")
+            vec_one = tvm.const(1, "int16x16")
+            pair_reduction_0 = tvm.call_llvm_intrin('int16x16',
+                                                  'llvm.x86.avx2.pmadd.ub.sw',
```
Comment (anijain2305) [inline topi/python/topi/x86/tensor_intrin.py]: Align (and other places as well)

File: topi/python/topi/x86/tensor_intrin.py [hunk 2]
```diff
@@ -275,3 +275,97 @@ def _instr(index):
...
+                                                  tvm.const(0, 'uint32'),
+                                                  vec_a, vec_b_1)
+            quad_reduction_1 = tvm.call_llvm_intrin('int32x8',
+                                                  'llvm.x86.avx2.pmadd.wd',
+                                                  tvm.const(0, 'uint32'),
+                                                  pair_reduction_1, vec_one)
+            if index == 0:
+                ib.emit(outs[0].vstore([0], quad_reduction_0))
+                ib.emit(outs[0].vstore([8], quad_reduction_1))
```
Comment (anijain2305) [inline topi/python/topi/x86/tensor_intrin.py]: Why have 2 quad_reductions in assembly instead of unrolling in the schedule?

Comment (anijain2305) [reply topi/python/topi/x86/tensor_intrin.py]: Oh maybe, you would have then change num_lanes etc

If thats tha case, @yzhliu is working on making it an argument. Maybe we can use that to simplify this.

Comment (kimishpatel) [reply topi/python/topi/x86/tensor_intrin.py]: @anijain2305. Correct. I wanted to be able to use the same memory layout which can be tensorized for avx2 and avx512.

**Comment by kimishpatel** [comment]
@anijain2305, thats a good question. I havent checked the performance of this agains fp32 model we are interested in. I will report the numbers soon. But we do get a lot of speed up on skylake with avx512 and assuming we dont get anything from AVX2 instructions itself due to latency issues you mention, there should still be some perf left from the memory bw persepctive.

**Comment by anijain2305** [comment]
> @anijain2305, thats a good question. I havent checked the performance of this agains fp32 model we are interested in. I will report the numbers soon. But we do get a lot of speed up on skylake with avx512 and assuming we dont get anything from AVX2 instructions itself due to latency issues you mention, there should still be some perf left from the memory bw persepctive.

Ok. This might be useful to do some paper calculations to see if we can get performance theoretically - https://www.agner.org/optimize/instruction_tables.pdf

**Comment by kimishpatel** [comment]
@anijain2305, do you mind if I land this? I would like to use it for some of the stuff we are working on.

File: tests/python/contrib/test_gemm_avx2_acc32.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+def test_avx2_int8_gemm_acc32():
+    m = 1024
+    n = 1024
+    k = 1024
+
+    X = tvm.placeholder((m, k), name='X', dtype="uint8")
+    W = tvm.placeholder((n, k), name='W', dtype="int8")
+
+    #peak = 280 // This needs measurement and description of what this number is for avx2 machine.
```
Comment (anijain2305) [inline tests/python/contrib/test_gemm_avx2_acc32.py]: Please remove the comments

**Comment by anijain2305** [comment]
> @anijain2305, do you mind if I land this? I would like to use it for some of the stuff we are working on.

Yes, please go ahead. The changes LGTM. Would appreciate if later you could update with the performance comparison.

**Comment by kimishpatel** [comment]
Thanks @anijain2305, will do so.

**Comment by kimishpatel** [comment]
Just as FYI. CI is failing on the test I added as it is dependent on this PR: https://github.com/dmlc/tvm/pull/3981/. I need to land that.

**Comment by kimishpatel** [comment]
@anijain2305, would mind merging this please? I dont think I can, plus it seems that my lint changes and addressing your comments might have voided your approval.

**Comment by anijain2305** [comment]
I don't have merge permissions :)
@tqchen can you please take a look?

**Comment by tqchen** [comment]
Thanks @kimishpatel @anijain2305

**Pull request merged**